### PR TITLE
runner: avoid using node's store in SetupState

### PR DESCRIPTION
**Pull request by bluetech**
This is a small continuation to #8219 which I noticed after the fact.

SetupState maintains its own state, so it can store the exception itself, instead of using the node's store, which is better avoided when possible.

This also reduces the lifetime of the reference-cycle-inducing exception objects which is never a bad thing.

### CODE REVIEW

File: src/_pytest/runner.py [hunk 1]
```diff
@@ -467,29 +466,33 @@ class SetupState:
...
 
     def __init__(self) -> None:
-        # Maps node -> the node's finalizers.
         # The stack is in the dict insertion order.
-        self.stack: Dict[Node, List[Callable[[], object]]] = {}
-
-    _prepare_exc_key = StoreKey[Union[OutcomeException, Exception]]()
+        self.stack: Dict[
+            Node,
+            Tuple[
```
Comment (RonnyPfannschmidt) [inline src/_pytest/runner.py:472]: Wouldn't this be a case for typing.NamedTuple

**Review by RonnyPfannschmidt — APPROVED** [review]
Lovely enhancement

File: src/_pytest/runner.py (hunk 1 above)
Comment (bluetech) [reply src/_pytest/runner.py:472]: I personally prefer it inline, that way everything is localized inside the class which makes it easier to follow for me. But I can definitely change it (probably to an attrs class) if you think it'd be better - let me know.

Comment (RonnyPfannschmidt) [reply src/_pytest/runner.py:472]: I agree with your assessment as we only use it in one place

**Review by nicoddemus — APPROVED** [review]

**Pull request merged**